File: resources/play/stubs/_lawp_player.py
import json
from pathlib import Path

LOG = Path.cwd() / "play-log.json"
DIRS = ("north", "south", "east", "west")


def _read():
    if LOG.exists():
        try:
            return json.loads(LOG.read_text(encoding="utf-8"))
        except json.JSONDecodeError:
            return []
    return []


def _write(row):
    data = _read()
    data.append(row)
    LOG.write_text(json.dumps(data), encoding="utf-8")


class _Player:
    def move(self, dir):
        if dir not in DIRS:
            _write({"op": "fault", "message": "move dir must be north, south, east, or west"})
            return
        _write({"op": "move", "dir": dir})

    def rotate(self, deg):
        if not isinstance(deg, (int, float)) or deg % 90 != 0:
            _write({"op": "fault", "message": "rotate must be a multiple of 90"})
            return
        _write({"op": "rotate", "deg": deg})

    def scale(self, n):
        if not isinstance(n, (int, float)):
            _write({"op": "fault", "message": "scale needs a number"})
            return
        _write({"op": "scale", "n": n})

    def say(self, text):
        _write({"op": "say", "text": str(text)})

    def wait(self, ticks):
        if not isinstance(ticks, int) or ticks < 0:
            _write({"op": "fault", "message": "wait ticks must be a whole number ≥ 0"})
            return
        _write({"op": "wait", "ticks": ticks})

    def __getattr__(self, name):
        def unknown(*_a, **_k):
            _write({"op": "fault", "message": f"unknown method {name}"})

        return unknown


Player = _Player()

File: resources/play/stubs/_lawp_player.mjs
import fs from 'node:fs'
import path from 'node:path'

const LOG = path.join(process.cwd(), 'play-log.json')

function readLog() {
  try {
    return JSON.parse(fs.readFileSync(LOG, 'utf8'))
  } catch {
    return []
  }
}

function write(row) {
  const list = readLog()
  list.push(row)
  fs.writeFileSync(LOG, JSON.stringify(list), 'utf8')
}

const DIRS = ['north', 'south', 'east', 'west']

const api = {
  move(dir) {
    if (!DIRS.includes(dir)) {
      write({ op: 'fault', message: 'move dir must be north, south, east, or west' })
      console.log('Player.move: bad dir')
      return
    }
    write({ op: 'move', dir })
    console.log('Player.move(' + dir + ')')
  },
  rotate(deg) {
    if (typeof deg !== 'number' || deg % 90 !== 0) {
      write({ op: 'fault', message: 'rotate must be a multiple of 90' })
      console.log('Player.rotate: bad deg')
      return
    }
    write({ op: 'rotate', deg })
    console.log('Player.rotate(' + deg + ')')
  },
  scale(n) {
    if (typeof n !== 'number') {
      write({ op: 'fault', message: 'scale needs a number' })
      return
    }
    write({ op: 'scale', n })
    console.log('Player.scale(' + n + ')')
  },
  say(text) {
    write({ op: 'say', text: String(text) })
    console.log('Player.say')
  },
  wait(ticks) {
    if (!Number.isInteger(ticks) || ticks < 0) {
      write({ op: 'fault', message: 'wait ticks must be a whole number ≥ 0' })
      console.log('Player.wait: bad ticks')
      return Promise.reject(new TypeError('wait ticks must be a whole number ≥ 0'))
    }
    const ms = Math.min(2000, ticks * 25)
    return new Promise((resolve) => {
      setTimeout(() => {
        write({ op: 'wait', ticks })
        console.log('Player.wait(' + ticks + ')')
        resolve()
      }, ms)
    })
  }
}

export const Player = new Proxy(api, {
  get(target, prop) {
    if (prop in target) return target[prop]
    return function unknown() {
      write({ op: 'fault', message: 'unknown method ' + String(prop) })
      console.log('Player.' + String(prop) + ': unknown')
    }
  }
})
